### ISSUE TITLE
Skip throttling for certain users

### ISSUE THREAD

**Issue by viktorsmari** [fablabbcn/smartcitizen-api#190]
We now throttle 100 req / minute per IP.

@oscgonfer asked to skip throttling on some users?

Options:
- Skip throttle if user `is_admin?` - What if we need to skip for Airflow users? Not all of them are admins?
- Skip throttle by supplying a `HEADER` when posting?
- Only throttle specific endpoints. Do we need to throttle all endpoints?

Example skipping if a special `HEADER` is supplied:

```ruby
# config/initializer/rack_attack.rb
safelist('do not throttle secret headers') do |req|
  req.env["HTTP_SKIP_THROTTLE"] == ENV['SKIP_THROTTLE']
end
```

`curl 'http://localhost:3000/devices' --header 'SKIP_THROTTLE:ABCD`

**Comment by pral2a** [comment]
I will suggest once you use a token throttle is per token, not IP. That can be problematic in workshops, hackathons, etc where many people is under the same NAT.

Then I will suggest we could simply remove throttling for researchers and admins.

Later on we could add a more generous throttling to avoid researchers to unintended brake the platform while testing code.

**Comment by oscgonfer** [comment]
Would this then work by supplying a certain API Oauth Bearer in the http request? From my point of view, that's the most comfortable solution.

**Comment by pral2a** [comment]
@oscgonfer yes, of course, but not a custom one, the Bearer Token Authorise and Authenticates you as a particular user, the platform stores for each user a role, currently: admin, researcher and citizen.

**Comment by pral2a** [comment]
@oscgonfer yes, of course, but not a custom one, the Bearer Token Authorize and Authenticates you as a particular user, the platform stores for each user a role (on a scale from 1-5)  we currently map to: admin, researcher and citizen. Then we can filter actions and assign particular behaviours based on that roles

https://github.com/fablabbcn/smartcitizen-api/blob/d3955403bb917355ebea9f35fdfa5b8ca28bf22a/app/models/user.rb#L100-L103

**Comment by timcowlishaw** [comment]
Hey folks - quick update on this - it's a little more complicated than I expected - due to the fact that the throttling happens in a Rack middleware, but authentication happens in the controller - therefore the current_user is not available at the time that rack::attack performs the request throttling (nor have we validated the credentials provided. 

I've tried to fix this by moving authentication up the stack to a new middleware, before the rack::attack one, but this is causing a whole load of other weird problems, so needs  a little more investigation.

Most of the errors i'm seeing are tests that expect `403 Forbidden` receiving a `401 Unauthorized` - however, many of these are for authentication errors (eg missing or incorrect tokens), which means a `401` might be more appropriate? if changing the expected response code in these situations is acceptable, then the matter might be settled already - if not, I need to look into alternatives to Rack::Attack which perform throttling from within a rails controller, which may be a little tricky as we'll need to make sure it's applied across all endpoints, but also happens after authentication.
